File: src/test/java/org/example/kalkulationsprogramm/service/BestellungPdfServiceTest.java
package org.example.kalkulationsprogramm.service;

import org.apache.pdfbox.Loader;
import org.apache.pdfbox.text.PDFTextStripper;
import org.example.kalkulationsprogramm.dto.Bestellung.BestellungResponseDto;
import org.junit.jupiter.api.Test;
import org.example.kalkulationsprogramm.repository.SchnittbilderRepository;
import org.mockito.Mockito;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class BestellungPdfServiceTest {

    @Test
    void createsPdfFile() throws Exception {
        BestellungService bestellungService = Mockito.mock(BestellungService.class);
        BestellungResponseDto dto = new BestellungResponseDto();
        dto.setLieferantId(1L);
        dto.setProjektName("Projekt A");
        dto.setExterneArtikelnummer("123");
        dto.setProduktname("Produkt");
        dto.setProdukttext("Text");
        dto.setKommentar("Kommentar");
        dto.setMenge(java.math.BigDecimal.ONE);
        dto.setEinheit("Stk");
        Mockito.when(bestellungService.findeOffeneBestellungen()).thenReturn(List.of(dto));

        SchnittbilderRepository schnittbilderRepository = Mockito.mock(SchnittbilderRepository.class);
        DateiSpeicherService dateiSpeicherService = Mockito.mock(DateiSpeicherService.class);
        FirmeninformationService firmeninformationService = Mockito.mock(FirmeninformationService.class);
        Mockito.when(firmeninformationService.loadLogoImage()).thenReturn(null);
        BestellungPdfService service = new BestellungPdfService(bestellungService, schnittbilderRepository, dateiSpeicherService, firmeninformationService);
        Path pdf = service.generatePdfForLieferant(1L);
        assertTrue(Files.size(pdf) > 0);
        String content = pdfText(pdf);
        assertTrue(content.contains("Bauvorhaben:"));
        assertTrue(content.contains("Rechnungen separat pro Auftrag"));
        assertTrue(endsAtPdfEof(pdf), "PDF enthält keine nachträglich angehängten Textbytes");
        Files.deleteIfExists(pdf);
    }

    @Test
    void createsPdfFileForUnknownLieferant() throws Exception {
        BestellungService bestellungService = Mockito.mock(BestellungService.class);
        BestellungResponseDto dto = new BestellungResponseDto();
        dto.setLieferantId(null);
        dto.setProjektName("Projekt B");
        dto.setExterneArtikelnummer("456");
        dto.setProduktname("ProduktB");
        dto.setProdukttext("TextB");
        dto.setKommentar("KommentarB");
        dto.setMenge(java.math.BigDecimal.ONE);
        dto.setEinheit("Stk");
        Mockito.when(bestellungService.findeOffeneBestellungen()).thenReturn(List.of(dto));

        SchnittbilderRepository schnittbilderRepository = Mockito.mock(SchnittbilderRepository.class);
        DateiSpeicherService dateiSpeicherService = Mockito.mock(DateiSpeicherService.class);
        FirmeninformationService firmeninformationService = Mockito.mock(FirmeninformationService.class);
        Mockito.when(firmeninformationService.loadLogoImage()).thenReturn(null);
        BestellungPdfService service = new BestellungPdfService(bestellungService, schnittbilderRepository, dateiSpeicherService, firmeninformationService);
        Path pdf = service.generatePdfForLieferant(null);
        assertTrue(Files.size(pdf) > 0);
        assertTrue(endsAtPdfEof(pdf), "PDF enthält keine nachträglich angehängten Textbytes");
        Files.deleteIfExists(pdf);
    }

    @Test
    void createsPdfForProjekt() throws Exception {
        BestellungService bestellungService = Mockito.mock(BestellungService.class);
        BestellungResponseDto dto = new BestellungResponseDto();
        dto.setProjektId(7L);
        dto.setProjektName("Projekt C");
        dto.setRootKategorieId(1);
        dto.setExterneArtikelnummer("789");
        dto.setProduktname("ProdC");
        dto.setProdukttext("TextC");
        dto.setKommentar("KommentarC");
        dto.setMenge(java.math.BigDecimal.ONE);
        dto.setEinheit("Stk");
        Mockito.when(bestellungService.findeOffeneBestellungen()).thenReturn(List.of(dto));

        SchnittbilderRepository schnittbilderRepository = Mockito.mock(SchnittbilderRepository.class);
        DateiSpeicherService dateiSpeicherService = Mockito.mock(DateiSpeicherService.class);
        FirmeninformationService firmeninformationService = Mockito.mock(FirmeninformationService.class);
        Mockito.when(firmeninformationService.loadLogoImage()).thenReturn(null);
        BestellungPdfService service = new BestellungPdfService(bestellungService, schnittbilderRepository, dateiSpeicherService, firmeninformationService);
        Path pdf = service.generatePdfForProjekt(7L);
        assertTrue(Files.size(pdf) > 0);
        assertTrue(endsAtPdfEof(pdf), "PDF enthält keine nachträglich angehängten Textbytes");
        Files.deleteIfExists(pdf);
    }

    private String pdfText(Path pdf) throws Exception {
        try (var document = Loader.loadPDF(Files.readAllBytes(pdf))) {
            return new PDFTextStripper().getText(document);
        }
    }

    private boolean endsAtPdfEof(Path pdf) throws Exception {
        String bytes = Files.readString(pdf, StandardCharsets.ISO_8859_1).stripTrailing();
        return bytes.endsWith("%%EOF");
    }
}

File: src/test/kotlin/org/example/kalkulationsprogramm/service/BestellungPdfServiceTest.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.dto.Bestellung.BestellungResponseDto
import org.example.kalkulationsprogramm.repository.SchnittbilderRepository
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import java.math.BigDecimal
import java.nio.charset.StandardCharsets
import java.nio.file.Files

class BestellungPdfServiceTest {

    @Test
    fun createsPdfFile() {
        val bestellungService = mock(BestellungService::class.java)
        val dto = BestellungResponseDto().apply {
            lieferantId = 1L
            projektName = "Projekt A"
            externeArtikelnummer = "123"
            produktname = "Produkt"
            produkttext = "Text"
            kommentar = "Kommentar"
            menge = BigDecimal.ONE
            einheit = "Stk"
        }
        `when`(bestellungService.findeOffeneBestellungen()).thenReturn(listOf(dto))

        val schnittbilderRepository = mock(SchnittbilderRepository::class.java)
        val dateiSpeicherService = mock(DateiSpeicherService::class.java)
        val firmeninformationService = mock(FirmeninformationService::class.java)
        `when`(firmeninformationService.loadLogoImage()).thenReturn(null)
        val service = BestellungPdfService(
            bestellungService,
            schnittbilderRepository,
            dateiSpeicherService,
            firmeninformationService,
        )

        val pdf = service.generatePdfForLieferant(1L)
        assertTrue(Files.size(pdf) > 0)
        val content = Files.readString(pdf, StandardCharsets.ISO_8859_1)
        assertTrue(content.contains("Bauvorhaben:"))
        assertTrue(content.contains("Rechnungen separat pro Auftrag"))
        Files.deleteIfExists(pdf)
    }

    @Test
    fun createsPdfFileForUnknownLieferant() {
        val bestellungService = mock(BestellungService::class.java)
        val dto = BestellungResponseDto().apply {
            lieferantId = null
            projektName = "Projekt B"
            externeArtikelnummer = "456"
            produktname = "ProduktB"
            produkttext = "TextB"
            kommentar = "KommentarB"
            menge = BigDecimal.ONE
            einheit = "Stk"
        }
        `when`(bestellungService.findeOffeneBestellungen()).thenReturn(listOf(dto))

        val schnittbilderRepository = mock(SchnittbilderRepository::class.java)
        val dateiSpeicherService = mock(DateiSpeicherService::class.java)
        val firmeninformationService = mock(FirmeninformationService::class.java)
        `when`(firmeninformationService.loadLogoImage()).thenReturn(null)
        val service = BestellungPdfService(
            bestellungService,
            schnittbilderRepository,
            dateiSpeicherService,
            firmeninformationService,
        )

        val pdf = service.generatePdfForLieferant(null)
        assertTrue(Files.size(pdf) > 0)
        Files.deleteIfExists(pdf)
    }

    @Test
    fun createsPdfForProjekt() {
        val bestellungService = mock(BestellungService::class.java)
        val dto = BestellungResponseDto().apply {
            projektId = 7L
            projektName = "Projekt C"
            rootKategorieId = 1
            externeArtikelnummer = "789"
            produktname = "ProdC"
            produkttext = "TextC"
            kommentar = "KommentarC"
            menge = BigDecimal.ONE
            einheit = "Stk"
        }
        `when`(bestellungService.findeOffeneBestellungen()).thenReturn(listOf(dto))

        val schnittbilderRepository = mock(SchnittbilderRepository::class.java)
        val dateiSpeicherService = mock(DateiSpeicherService::class.java)
        val firmeninformationService = mock(FirmeninformationService::class.java)
        `when`(firmeninformationService.loadLogoImage()).thenReturn(null)
        val service = BestellungPdfService(
            bestellungService,
            schnittbilderRepository,
            dateiSpeicherService,
            firmeninformationService,
        )

        val pdf = service.generatePdfForProjekt(7L)
        assertTrue(Files.size(pdf) > 0)
        Files.deleteIfExists(pdf)
    }
}
